XenAdmin/SettingsPanels/CPUMemoryEditPage.cs

```diff
@@ -220,39 +214,106 @@ public void Repopulate()
 
-            _ValidToSave = true;
+            if (isVcpuHotplugSupported)
+                PopulateVCPUsAtStartup(_OrigVCPUsMax, _OrigVCPUsAtStartup);
+        }
+
+        private void PopulateVCPUComboBox(ComboBox comboBox, long min, long max, long currentValue, Predicate<long> isValid)
+        {
+            comboBox.BeginUpdate();
+            comboBox.Items.Clear();
+            for (long i = min; i <= max; ++i)
+            {
+                if (i == currentValue || isValid(i))
+                    comboBox.Items.Add(i);
+            }
+            if (currentValue > max)
+                comboBox.Items.Add(currentValue);
+            comboBox.SelectedItem = currentValue;
+            comboBox.EndUpdate();
         }
 
         private void PopulateVCPUs(long maxVCPUs, long currentVCPUs)
         {
-            comboBoxVCPUs.BeginUpdate();
-            comboBoxVCPUs.Items.Clear();
-            for (long i = 1; i <= maxVCPUs; ++i)
+            PopulateVCPUComboBox(comboBoxVCPUs, 1, maxVCPUs, currentVCPUs, i => comboBoxTopology.IsValidVCPU(i));
+        }
+
+        private void PopulateVCPUsAtStartup(long max, long currentValue)
+        {
+            long min = vm.power_state == vm_power_state.Halted ? 1 : _OrigVCPUsAtStartup;
+            PopulateVCPUComboBox(comboBoxInitialVCPUs, min, max, currentValue, i => true);
+        }
+
+        private string GetRubric()
+        {
+            StringBuilder sb = new StringBuilder();
+            sb.Append(Messages.VM_CPUMEMPAGE_RUBRIC);
+            // add hotplug text
+            if (isVcpuHotplugSupported)
+                sb.Append(Messages.VM_CPUMEMPAGE_RUBRIC_HOTPLUG);
+            // add power state warning
+            if (vm.power_state != vm_power_state.Halted)
+            {
+                sb.AppendLine();
+                sb.AppendLine();
+                sb.Append(isVcpuHotplugSupported ? Messages.VM_CPUMEMPAGE_MAX_VCPUS_READONLY : Messages.VCPU_ONLY_WHEN_HALTED);
+            }
+            // add power state warning for Current number of vCPUs
+            if (isVcpuHotplugSupported && vm.power_state != vm_power_state.Halted && vm.power_state != vm_power_state.Running)
             {
-                if (i == currentVCPUs || comboBoxTopology.IsValidVCPU(i))
-                    comboBoxVCPUs.Items.Add(i);
+                sb.Append(isVcpuHotplugSupported ? Messages.VM_CPUMEMPAGE_CURRENT_VCPUS_READONLY : Messages.VCPU_ONLY_WHEN_HALTED);
             }
-            if (currentVCPUs > maxVCPUs)
-                comboBoxVCPUs.Items.Add(currentVCPUs);
-            comboBoxVCPUs.SelectedItem = currentVCPUs;
-            comboBoxVCPUs.EndUpdate();
+            return sb.ToString();
         }
 
         public bool HasChanged
         {
-            get { return HasVCPUChanged || HasMemoryChanged || HasTopologyChanged; }
+            get { return HasVCPUChanged || HasMemoryChanged || HasTopologyChanged || HasVCPUsAtStartupChanged; }

```

Comment:
I think it's confusing that HasVCPUChanged includes the VCPU-max and the weight, but the HasVCPUsAtStartupChanged is separate. Maybe they should just all be separate.